crates/ruma-client-api/src/discovery/discover_homeserver.rs

```diff
@@ -112,3 +133,187 @@ impl TileServerInfo {
+    pub fn livekit(service_url: String) -> Self {
+        Self::LiveKit(LiveKitRtcFocusInfo { service_url })
+    }
+
+    /// Returns a reference to the focus type of this RTC focus.
+    pub fn focus_type(&self) -> &str {
+        match self {
+            Self::LiveKit(_) => "livekit",
+            Self::_Custom(custom) => &custom.focus_type,
+        }
+    }
+
+    /// Returns the associated data.
+    ///
+    /// The returned JSON object won't contain the `focus_type` field, please use
+    /// [`.focus_type()`][Self::focus_type] to access that.
+    ///
+    /// Prefer to use the public variants of `RtcFocusInfo` where possible; this method is meant to
+    /// be used for custom focus types only.
+    pub fn data(&self) -> Cow<'_, JsonObject> {
+        fn serialize<T: Serialize>(object: &T) -> JsonObject {
+            match serde_json::to_value(object).expect("rtc focus type serialization to succeed") {
+                JsonValue::Object(object) => object,
+                _ => panic!("all rtc focus types must serialize to objects"),
+            }
+        }
+
+        match self {
+            Self::LiveKit(info) => Cow::Owned(serialize(info)),
+            Self::_Custom(info) => Cow::Borrowed(&info.data),
+        }
+    }
+}
+
+/// Information about a LiveKit RTC focus.
+#[cfg(feature = "unstable-msc4143")]
+#[derive(Clone, Debug, Deserialize, Serialize)]
+#[cfg_attr(not(ruma_unstable_exhaustive_types), non_exhaustive)]
+pub struct LiveKitRtcFocusInfo {
+    /// The service URL for the LiveKit service.
+    #[serde(rename = "livekit_service_url")]
+    pub service_url: String,
+}
+
+/// Information about a custom RTC focus type.
+#[doc(hidden)]
+#[cfg(feature = "unstable-msc4143")]
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct CustomRtcFocusInfo {
+    /// The type of RTC focus.
+    #[serde(rename = "type")]
+    focus_type: String,
+
+    /// Remaining RTC focus data.
+    #[serde(flatten)]
+    data: JsonObject,
+}
+
+#[cfg(test)]
+mod tests {
+    #[cfg(feature = "unstable-msc4143")]
+    use serde_json::{from_value as from_json_value, json, to_value as to_json_value};
+
+    #[cfg(feature = "unstable-msc4143")]
+    use super::RtcFocusInfo;
+
+    #[test]
+    #[cfg(feature = "unstable-msc4143")]
+    fn test_livekit_rtc_focus_deserialization() {
+        // Given the JSON for a LiveKit RTC focus.
+        let json = json!({
+            "type": "livekit",
+            "livekit_service_url": "https://livekit.example.com"
+        });
+
+        // When deserializing it into an RtcFocusInfo.
+        let focus: RtcFocusInfo = from_json_value(json).unwrap();
+
+        // Then it should be recognized as a LiveKit focus with the correct service URL.
+        match focus {
+            RtcFocusInfo::LiveKit(info) => {
+                assert_eq!(info.service_url, "https://livekit.example.com");
+            }
+            RtcFocusInfo::_Custom(_) => panic!("Expected LiveKit focus"),
+        }
```

Comment:
Sorry I didn't catch that before, here we would use `assert_matches2::assert_matches`:

```suggestion
        assert_matches!(focus, RtcFocusInfo::LiveKit(info));
        assert_eq!(info.service_url, "https://livekit.example.com");
```